src/app.md

```diff
@@ -725,6 +726,80 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[ex-semver-latest]: #ex-semver-latest
+<a name="ex-semver-latest"></a>
+## Find the latest version satisfying given range
+[![semver-badge]][semver] [![cat-config-badge]][cat-config]
+
+Given a list of version &strs, finds the latest [`semver::Version`] that satisfying a given [`semver::VersionReq`] using [`VersionReq::matches`].
+
+```rust
+# #[macro_use]
+# extern crate error_chain;
+extern crate semver;
+
+use semver::{AlphaNumeric, Numeric, Version, VersionReq};
+use std::vec::IntoIter;
+#
+# error_chain! {
+#     foreign_links {
+#         SemVer(semver::ReqParseError);
+#     }
+# }
+
+fn find_max_matching_version(
```

Comment:
If you go with signature such as this one:

```rust
fn find_max_matching_version<'a, I>(
    version_req_str: &str,
    iterable: I,
) -> Result<Option<Version>>
where
    I: IntoIterator<Item = &'a str>,
{
```

Then you will be able to omit both `into_iter()` calls below and 

also rustfmt allows us to put whole function declaration in a single line so lets do it for readability.

Reply:
When I run the code with this signature, I get this error:

```
error[E0599]: no method named `filter_map` found for type `I` in the current scope
  --> src/main.rs:33:10
   |
33 |         .filter_map(|s| Version::parse(s).ok())
   |          ^^^^^^^^^^
   |
   = note: the method `filter_map` exists but the following trait bounds were not satisfied:
           `&mut I : std::iter::Iterator`
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `filter_map`, perhaps you need to implement it:
           candidate #1: `std::iter::Iterator`
```
I couldn't figure out how to resolve this compiler error.  Do I need to implement filter_map for I?

Reply:
@anna-liao , have you tried to remove the two below `into_iter()` calls (in the asserts) and go `iterable.into_iter().filter_map(...` in `find_max_matching_version` function?

Reply:
> Do I need to implemDo I need to implement filter_map for I?

Nope. Actually the error message could be a lot better here.

Let me try to explain what is going on here. `IntoIterator` is a trait bound here making the `find_max_matching_version`  function generic over type implementing the trait. So it will accept now things like `Vec`, `HashSet` (and almost any other container) or any `Iterator` as long as their `Item` is `&str`.

But while all these types are accepted as the parameter, they are not an iterator yet (which itself implement `filter_map`). You will have to actually use the `into_iter` trait method to turn the argument into an `Iterator`.

This is how almost all conversion trait bounds are used (`AsRef<T>`,  `Into<T>`, etc). You use the trait bound to accept a range of types and use the trait method inside the function to get parameter converted into the type you want. I hope this helps :)

Reply:
@ludwigpacifici @budziq Ah thanks for the explanation!  I have tested the code as suggested,  and updated the PR